### PR TITLE
*: Fix a DDL bug about wait time

### PR DESCRIPTION
**Pull request by zimulala**
If the DDL job is done, this job will put to the DDL history queue,  then the owner waits for 2*lease or checks all servers schema version. 
At the same time, the TiDB that puts the DDL job to the DDL queue (the TiDB that receives the DDL statement) may get the DDL job from history queue, and it will think that this statement has been completed.
So it is a bug, this PR fixed the bug.

### CODE REVIEW

File: ddl/ddl_test.go [hunk 1]
```diff
@@ -94,14 +98,6 @@ func checkHistoryJobArgs(c *C, ctx context.Context, id int64, args *historyJobAr
 	if args.db != nil && len(args.tblIDs) == 0 {
 		return
 	}
-	// only for dropping schema job
```
Comment (shenli) [inline ddl/ddl_test.go]: Why remove this?

File: model/ddl.go [hunk 2]
```diff
@@ -186,6 +189,16 @@ func (job *Job) IsFinished() bool {
 	return job.State == JobDone || job.State == JobRollbackDone || job.State == JobCancelled
 }
 
+// IsCancelled returns whether the job is cancelled or not.
+func (job *Job) IsCancelled() bool {
+	return job.State == JobCancelled || job.State == JobRollbackDone
+}
+
+// IsSynced returns whether the job is waited or not.
```
Comment (shenli) [inline model/ddl.go]: whether the DDL modification is synced among all tidb-servers

File: model/ddl.go [hunk 3]
```diff
@@ -144,11 +144,14 @@ func (job *Job) GetRowCount() int64 {
 }
 
 // Encode encodes job with json format.
-func (job *Job) Encode() ([]byte, error) {
+// updateRawArgs is used to determine whether to update the raw args.
```
Comment (shenli) [inline model/ddl.go]: When we need to update raw args?

Comment (zimulala) [inline model/ddl.go]: The background job will need to update raw args.

File: ddl/ddl_test.go (hunk 1 above)
Comment (zimulala) [inline ddl/ddl_test.go]: Because the history job is DDL job, its RawArgs is covered by Args when doing encode. So it doesn't have the RawArgs value now.

**Comment by zimulala** [comment]
PTAL @shenli @tiancaiamao @disksing

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -214,10 +220,12 @@ func (d *ddl) handleDDLJobQueue() error {
 			schemaVer = d.runDDLJob(t, job)
 			if job.IsFinished() {
 				binloginfo.SetDDLBinlog(d.workerVars.BinlogClient, txn, job.ID, job.Query)
```
Comment (hanfei1991) [inline ddl/ddl_worker.go]: If job is cancelled, do we still need to set binlog ?

File: ddl/ddl_worker.go [hunk 5]
```diff
@@ -188,7 +188,7 @@ func (d *ddl) handleDDLJobQueue() error {
 				return errors.Trace(err)
 			}
 
-			if job.IsRunning() {
+			if job.IsRunning() || job.IsSynced() {
```
Comment (hanfei1991) [inline ddl/ddl_worker.go]: Why should we check if job is synced? if a job's state is synced, can we read it from ddl queue?

File: ddl/ddl_test.go [hunk 6]
```diff
@@ -75,6 +75,10 @@ func checkEqualTable(c *C, t1, t2 *model.TableInfo) {
 	c.Assert(t1.AutoIncID, DeepEquals, t2.AutoIncID)
 }
 
+func checkHistoryJob(c *C, job *model.Job) {
```
Comment (hanfei1991) [inline ddl/ddl_test.go]: Why extract a function to assert ? According the function's name, I cannot understand what will be checked.

File: ddl/ddl_worker.go (hunk 4 above)
Comment (zimulala) [inline ddl/ddl_worker.go]: This is old logic, I don't change it.

Comment (hanfei1991) [inline ddl/ddl_worker.go]: Will this old logic cause a bug?

File: model/ddl.go [hunk 7]
```diff
@@ -144,11 +144,14 @@ func (job *Job) GetRowCount() int64 {
...
+func (job *Job) Encode(updateRawArgs bool) ([]byte, error) {
 	var err error
-	job.RawArgs, err = json.Marshal(job.Args)
-	if err != nil {
-		return nil, errors.Trace(err)
+	if updateRawArgs {
```
Comment (hanfei1991) [inline model/ddl.go]: Why not just check if job.Args is nil ? When the job Args is not nil but we should not encode it?

File: ddl/ddl_worker.go (hunk 5 above)
Comment (zimulala) [inline ddl/ddl_worker.go]: We want to make sure the job waits for 2*lease or finds all servers' schema version are the latest version. You can read the comments of `JobSynced`.

File: model/ddl.go (hunk 7 above)
Comment (zimulala) [inline model/ddl.go]: Use this function more frequently, if there is a problem there will be bad positioning.

**Comment by hanfei1991** [comment]
LGTM

File: ddl/ddl_worker.go (hunk 4 above)
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: Cancel is not the old logic.

Comment (tiancaiamao) [inline ddl/ddl_worker.go]: Would you use the information in Job and will the `Cancel` status affect you?  @GregoryIan

Comment (GregoryIan) [inline ddl/ddl_worker.go]: wont affect me

Comment (zimulala) [inline ddl/ddl_worker.go]: @tiancaiamao Cancel is the old logic.  https://github.com/pingcap/tidb/pull/3498/files#diff-c87d71f8121c3ee6f07826da99de9eabR189

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**